src/town_cmd.cpp

```diff
@@ -517,6 +517,37 @@ static void MakeTownHouseBigger(TileIndex tile)
 	if (flags & BUILDING_HAS_4_TILES) MakeSingleHouseBigger(TILE_ADDXY(tile, 1, 1));
 }
 
+/**
+ * Generate cargo for a house, scaled by the current economy scale.
+ * @param t The current town.
+ * @param ct Type of cargo to generate, usually CT_PASSENGERS or CT_MAIL.
+ * @param amount The number of cargo units.
+ * @param stations Available stations for this house.
+ * @param affected_by_recession Is this cargo halved during recessions?
+ */
+static void TownGenerateCargo(Town *t, CargoID ct, uint amount, StationFinder &stations, bool affected_by_recession)
+{
+	uint cargo_amount = amount;
+
+	/* All production is halved during a recession (except for NewGRF-supplied town cargo). */
+	if (affected_by_recession && EconomyIsInRecession()) {
+		cargo_amount = (cargo_amount + 1) >> 1;
+	}
+
+	/* If we're not producing any cargo, we can bail out now. */
+	if (cargo_amount == 0) return;
+
+	/* Scale by cargo scale setting. */
+	cargo_amount = ScaleByCargoScale(cargo_amount);
+
+	/* Actually generate cargo and update town statistics. */
+	uint moved = MoveGoodsToStation(ct, cargo_amount, ST_TOWN, t->index, stations.GetStations());
+
+	const CargoSpec *cs = CargoSpec::Get(ct);
+	t->supplied[cs->Index()].new_max += cargo_amount;
+	t->supplied[cs->Index()].new_act += moved;
```

Comment:
Just to confirm, I just don't know this stuff well enough: CargoSpec of CT_PASSENGERS returns the same index as CT_PASSENGERS?

From a coding-point-of-view, that is different in the old code and new code. Old code does `t->supplied[CT_PASSENGERS]`, and new code first looks it up via `CargoSpec` to use `cs->Index()`. I have zero clue if that has any impact.

Reply:
Yes, CT_PASSENGERS is the index. This code is copied from the implementation in the house production callback, whereas CT_PASSENGERS is just the hardcoded version.

Reply:
There is no reason to look up the cargo spec and then use `cs->Index()`. `ct` is already that index.